### DISCUSSION TITLE
Ceph COSI Driver Image(v0.1.2) Update for Rook v1.15

### DISCUSSION THREAD

**Discussion by thotz** [rook/rook#14297] [General]
The ceph-cosi driver could support multiple ceph-cosi drivers in the k8s from v0.1.2 onwards. This will change the existing driver name from `ceph.objectstorage.k8s.io` to `rook-ceph.ceph.objectstorage.k8s.io`. And it impacts the existing cosi buckets. One way to resolve this update the existing `bucketclass` and `bucketaccessclass` crds with driver name

The existing users need to update driver name in `bucketclass` and `bucketaccessclass` post upgrading to Rook v1.15

from 

```yaml
kind: BucketClass
apiVersion: objectstorage.k8s.io/v1alpha1
metadata:
  name: sample-bcc
driverName: ceph.objectstorage.k8s.io
deletionPolicy: Delete
parameters:
  objectStoreUserSecretName: rook-ceph-object-user-my-store-cosi
  objectStoreUserSecretNamespace: rook-ceph
---
kind: BucketAccessClass
apiVersion: objectstorage.k8s.io/v1alpha1
metadata:
  name: sample-bac
driverName: ceph.objectstorage.k8s.io
authenticationType: KEY
parameters:
  objectStoreUserSecretName: rook-ceph-object-user-my-store-cosi
  objectStoreUserSecretNamespace: rook-ceph
```

to

```yaml
kind: BucketClass
apiVersion: objectstorage.k8s.io/v1alpha1
metadata:
  name: sample-bcc
driverName: rook-ceph.ceph.objectstorage.k8s.io
deletionPolicy: Delete
parameters:
  objectStoreUserSecretName: rook-ceph-object-user-my-store-cosi
  objectStoreUserSecretNamespace: rook-ceph
---
kind: BucketAccessClass
apiVersion: objectstorage.k8s.io/v1alpha1
metadata:
  name: sample-bac
driverName: rook-ceph.ceph.objectstorage.k8s.io
authenticationType: KEY
parameters:
  objectStoreUserSecretName: rook-ceph-object-user-my-store-cosi
  objectStoreUserSecretNamespace: rook-ceph
```

and perform
```bash
kubectl apply -f deploy/example/cosi/bucketclass.yaml -f deploy/example/cosi/bucketaccessclass.yaml
```

Now edit exisiting `bucket` CRDs  via
```
kubectl edit buckets.objectstorage.k8s.io <bucket name>

apiVersion: objectstorage.k8s.io/v1alpha1
kind: Bucket
metadata:
  creationTimestamp: "2024-06-13T11:41:09Z"
  finalizers:
  - cosi.objectstorage.k8s.io/bucket-protection
  - cosi.objectstorage.k8s.io/bucketaccess-bucket-protection
  generation: 1
  name: sample-bcc3c70893e-dc3b-4615-a62f-52662fe4b4f4
  resourceVersion: "6824"
  uid: 934e53dc-d50e-4e53-8ae9-27ef088684f0
spec:
  bucketClaim:
    name: sample-bucket
    namespace: default
    uid: 3c70893e-dc3b-4615-a62f-52662fe4b4f4
  bucketClassName: sample-bcc
  deletionPolicy: Delete
  driverName: ceph.objectstorage.k8s.io
  parameters:
    objectStoreUserSecretName: rook-ceph-object-user-my-store-cosi
    objectStoreUserSecretNamespace: rook-ceph
  protocols:
  - s3
status:
  bucketID: sample-bcc3c70893e-dc3b-4615-a62f-52662fe4b4f4
  bucketReady: true


```
And change `driverName` in the `spec` from `ceph.objectstorage.k8s.io` to `rook-ceph.ceph.objectstorage.k8s.io`

**Comment by BlaineEXE** [comment]
Will any existing Buckets, BucketClaims, and BucketAccesses be affected by the change? Do users need to update these resources?

**Reply by travisn** [reply]
Right, seems like these would be broken in some way. @thotz Please confirm when you have verified these upgrade steps.

**Reply by thotz** [reply]
@travisn : 
The driver information is present in `Buckets` as well, so the user need to modify existing bucket resources as well

I have created a cluster, cephcosi driver,  bucketclass, bucketclaim, bucketaccessclass, bucketaccess on it. Update the Rook Operator image to the latest this change . Modified resources bucketclass, bucketaccessclass and buckets after the update. I was able to create new buckets, bucket access. Also I was able to delete old buckets, bucketaccess etc.

**Reply by travisn** [reply]
Ok good to hear everything was working. It's just not intuitive to me. Do you understand why the existing buckets were working after the provisoiner name? If the provisioner name changed I just anticipated it would be a problem for those existing resources. At least with the csi driver it is expected to break if the provisioner name changes.

**Reply by thotz** [reply]
The storageclass won't able to modify like bucketclass/bucketaccessclass do(May it is a bug in COSI Spec). I don't know about pv/pvc at least I am guessing pv cannot be modified. So csi driver cannot fulfil the requests since driver name was different

**Reply by travisn** [reply]
@thotz Are you saying it's a bug that it works? I still don't understand.

**Reply by thotz** [reply]
@travisn I meant we cannot modify storageclass once created. But for bucketclass/bucketaccessclass, they are allowing to update as of now. I don't know whether it is intentional or not. Maybe it will change in later releases.  AFAIR the `Buckets` CRDs are intended to allow changes or modification.
